### DISCUSSION TITLE
Design pattern for selective initialization of ORMs

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7611] [Usage Questions]
Sometimes when you load objects from the database, you want them to participate completely in the ecosystem of the app. Other times, you are more interested in doing some introspection or otherwise using loaded information but without behaviors. In other words, there may be some instances of classA which you want to be "live" and others which you want to be zombies. The load() hook provides a mechanism for doing post-load work and passing a QueryContext. However, I have had a hard time finding how to pass contextual cookies that would allow load() handlers to build a zombie vs a non-zombie, for example. Do we have to rely on some global variable and that there can only be one query/load/fetch happening at once? What's the right pattern?

As always thanks for the help. I did look around but couldn't find the thread. I'm digging the with_polymorphic() functionality. I figure one way around this problem is to do some simple inflation from raw rows myself ...

**Comment by zzzeek** [comment]
i dont quite get what a "zombie" means.    you have a class, with attributes and methods on it, they have behaviors.  what are the differences in behaviors that a "zombie" would have?  I would assume you mean, when you access attributes on it, nothing happens, it doesnt try to load anything, or persist anything.  doesnt that just mean you want to load the objects and then detach them from the session?   i dont think that can work inside the load() event directly.

**Reply by aathan** [reply]
We could assume that the word "zombie" in this discussion refers to behaviors of the instantiated object that are not specific to SQLA. Maybe I'm going about things incorrectly on some fundamental level, but some of my classes perform various scaffolding tasks in `__init__()` and or `on_load()` (the latter being a function I register for load() events). In the semantic scope of those functions we don't know "why" the object is being loaded or initialized without passing in that information or making it available in some global scope (such as a contextvar, or per what motivated my original question, say the QueryContext available in the load() event).

In other words, I start some complex loading activity in one place of the program for some specific purpose X, then 42 call stacks deep somewhere else, an object's __init__() or on_load() ends up running. When that object scaffolds itself into the program, because it is being loaded for the purpose X, it should do certain things differently. For example, an Order object in a trading system shouldn't register itself for live eventing if it is being loaded for historical data analysis.

So, I was wondering if there is a canonical way to pass contextual information to objects being inflated from the DB. Another way I could imagine this could happen is by attaching attributes to the session into which an object is being loaded and then (not sure the APIs, hint appreciated) modulating initialization behavior based on the session attributes -- E.g., this is an "analytical" Session vs a "live" Session.

(yes lets assume an async/gevent/threaded program where there may be multiple DB related activities happening "at once").

**Reply by aathan** [reply]
While I have your attention, is there a utility function that, given a row containing the `polymorphic_on` column of an inheritance hierarchy would give you the class that should be instantiated from that row? I've rolled my own, but surely that's something SQLA must have internally (and its implementation is probably more efficient than mine).

**Reply by zzzeek** [reply]
> In other words, I start some complex loading activity in one place of the program for some specific purpose X, then 42 call stacks deep somewhere else, an object's **init**() or on_load() ends up running. When that object scaffolds itself into the program, because it is being loaded for the purpose X, it should do certain things differently. For example, an Order object in a trading system shouldn't register itself for live eventing if it is being loaded for historical data analysis.

well from my POV i dont think an Order object should be doing such a thing upon load in any case.   I'd be loading that object from some function that would run the results through a filter that then operates upon the objects to establish them within the desired contexts.  This doesnt mean Order cant be the object that knows how to register itself, it's just that the outside caller would be passing a message to Order after it's outside of the load process (e.g. something like Order.establish_for_context(Context.LOADED_AND_NEEDS_LIVE_EVENTING) ) .   This is "safer" because within the "load" process, other collections and related attributes are not fully loaded yet, and it's best to avoid reentrant "loading" inside that event.    

but that doesnt mean you can't do it the way you describe,  if you want to send messages to objects that are being loaded which they can use to make decisions within the load() or refresh() event, there's at least three APIs which can be used for this.  One is to set up execution options with the execute call which can then be consumed from the QueryContext.  Another is to create [UserDefinedOption](https://github.com/sqlalchemy/sqlalchemy/blob/71248e34befe26f3c116dd52d782c519a5022168/lib/sqlalchemy/orm/interfaces.py#L842) classes and pass those through (this class seems to be missing from published API documentation, but you can see an example of them at https://docs.sqlalchemy.org/en/14/_modules/examples/dogpile_caching/caching_query.html ).  The UserDefinedOption object in particular has the ability to be carried along with an object and its related objects, such as you'd be able to see it on a subsequent lazyload.        Still another option as you mention is to put data into Session.info, and all the objects within a certain Session would be able to see that data by looking at the .info for their Session.

> 
> 
> (yes lets assume an async/gevent/threaded program where there may be multiple DB related activities happening "at once").

this issue is not really impacted by that because Sessions / attached objects should remain local to a singlar particular thread/task/greenlet in any case.

**Reply by zzzeek** [reply]
> While I have your attention, is there a utility function that, given a row containing the `polymorphic_on` column of an inheritance hierarchy would give you the class that should be instantiated from that row? I've rolled my own, but surely that's something SQLA must have internally (and its implementation is probably more efficient than mine).

I'd assume you are looking in [Mapper.polymorphic_map](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=mapper%20polymorphic_map#sqlalchemy.orm.Mapper.polymorphic_map) right now, that is a dictionary that's public.   I would stick with that.   if you aren't using that and have some kind of class magic going on, then yeah just use that dictionary.

**Reply by aathan** [reply]
Sure yes. One can always loop through all query results to set the loaded objects into a particular mode after the load is complete. I guess we've covered several possible designs that are relevant. I agree that doing things "at init time" or "at load time" has its warts ... and sometimes you're also dealing with legacy code/complex systems where refactoring that stuff is a challenge. I was looking for ways to make the behavioral mods be more "declarative" in the querying call (without having to always pass queries through a custom querier than can do post-hoc setup). I'm also imagining that some load operations may have far-flung side effects that would benefit from knowing the purpose of the load (thus contextvar-ish solutions).

Anyway, I get your drift.

You state: "One is to set up execution options with the execute call which can then be consumed from the QueryContext". I'm guessing you mean an "options()"  as in for example "Query(...).options(raiseload('*'))`?

I am pretty sure QueryContext is basically opaque in the documentation, but I can go read the source. I gather you mean that it holds those options().

**Reply by zzzeek** [reply]
UserDefinedOption is meant to be a public API, and it is mostly intended to be used with the ORMExecuteState event structure, where you get it here: https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=ormexecutestate#sqlalchemy.orm.ORMExecuteState.user_defined_options

the idea is that you would put the information you want into UserDefinedOption objects, and you could then either post-process the result using the [do_orm_execute](https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event hook, or if you wanted to access the object within the load event it is accessible via the [load_options](https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4/lib/sqlalchemy/orm/state.py#L71) attribute of InstanceState, which at the moment is kind of quasi-public - InstanceState itself is what you get when you call [inspect()](https://docs.sqlalchemy.org/en/14/orm/internals.html?highlight=instancestate#sqlalchemy.orm.InstanceState) on any mapped object.

**Reply by aathan** [reply]
Ohhh ... you originally wrote "One is to set up execution options with the execute call... Another is ... UserDefinedOption ..." so I thought the execution options you were referring to are a different thing than UserDefinedOption. Got it. Thanks.

**Reply by zzzeek** [reply]
this is more adventurous territory with lots of new APIs and stuff, i think it's way better than what we had before but the full story is not fully written out in the docs as you've noted.
